MapboxGLAndroidSDK/src/cpp/style/sources/source.cpp

```diff
@@ -50,205 +55,263 @@ static std::unique_ptr<Source> createSourcePeer(jni::JNIEnv& env,
-            return jni::Box(env, jni::jint(delta.value()));
-        }
+jni::Local<jni::Integer> Source::getPrefetchZoomDelta(jni::JNIEnv &env) {
+    auto guard = source.lock();
+    if (!source) {
+        mbgl::Log::Error(mbgl::Event::JNI, "Failed to get source prefetch zoom delta: " + ERR_MSG_DEAD_CORE_PEER);
         return jni::Local<jni::Integer>(env, nullptr);
     }
+    auto delta = source->getPrefetchZoomDelta();
+    if (delta.has_value()) {
+        return jni::Box(env, jni::jint(delta.value()));
+    }
+    return jni::Local<jni::Integer>(env, nullptr);
+}
 
-    void Source::setMaxOverscaleFactorForParentTiles(jni::JNIEnv& env, jni::Integer& maxOverscaleFactor) {
-        if (!maxOverscaleFactor) {
-            source.setMaxOverscaleFactorForParentTiles(nullopt);
-        } else {
-            source.setMaxOverscaleFactorForParentTiles(jni::Unbox(env, maxOverscaleFactor));
-        }
+void
+Source::setMaxOverscaleFactorForParentTiles(jni::JNIEnv &env, jni::Integer &maxOverscaleFactor) {
+    auto guard = source.lock();
+    if (!source) {
+        mbgl::Log::Error(mbgl::Event::JNI, "Failed to set source max overscale factor for parent tiles: " + ERR_MSG_DEAD_CORE_PEER);
+        return;
+    }
+    if (!maxOverscaleFactor) {
+        source->setMaxOverscaleFactorForParentTiles(nullopt);
+    } else {
+        source->setMaxOverscaleFactorForParentTiles(jni::Unbox(env, maxOverscaleFactor));
     }
+}
 
-    jni::Local<jni::Integer> Source::getMaxOverscaleFactorForParentTiles(jni::JNIEnv& env) {
-        auto maxOverscaleFactor = source.getMaxOverscaleFactorForParentTiles();
-        if (maxOverscaleFactor) {
-            return jni::Box(env, jni::jint(*maxOverscaleFactor));
-        }
+jni::Local<jni::Integer> Source::getMaxOverscaleFactorForParentTiles(jni::JNIEnv &env) {
+    auto guard = source.lock();
+    if (!source) {
+        mbgl::Log::Error(mbgl::Event::JNI, "Failed to get source max overscale factor for parent tiles: " + ERR_MSG_DEAD_CORE_PEER);
         return jni::Local<jni::Integer>(env, nullptr);
     }
+    auto maxOverscaleFactor = source->getMaxOverscaleFactorForParentTiles();
+    if (maxOverscaleFactor) {
+        return jni::Box(env, jni::jint(*maxOverscaleFactor));
+    }
+    return jni::Local<jni::Integer>(env, nullptr);
+}
 
-    void Source::addToStyle(JNIEnv& env, const jni::Object<Source>& obj, mbgl::style::Style& style) {
-        if (!ownedSource) {
-            throw std::runtime_error("Cannot add source twice");
-        }
+void Source::addToStyle(JNIEnv &env, const jni::Object<Source> &obj, mbgl::style::Style &style) {
+    if (!ownedSource) {
+        throw std::runtime_error("Cannot add source twice");
+    }
+
+    // Add source to style and release ownership
+    style.addSource(std::move(ownedSource));
 
-        // Add source to style and release ownership
-        style.addSource(std::move(ownedSource));
+    // Add peer to core source
+    source->peer = this;
```

Comment:
source is weak_ptr, can be null, same for line https://github.com/mapbox/mapbox-gl-native-android/pull/432/files#diff-2c2b9946805e54eb4377357eca5763c3R207

Reply:
Fixed.